README.md
# [![bitmovin](https://cloudfront-prod.bitmovin.com/wp-content/themes/Bitmovin-V-0.1/images/logo3.png)](http://www.bitmovin.com)
[![Build Status](https://travis-ci.org/bitmovin/bitcodin-python.svg?branch=master)](https://travis-ci.org/bitmovin/bitcodin-python)
[![Coverage Status](https://coveralls.io/repos/bitmovin/bitcodin-python/badge.svg?branch=master)](https://coveralls.io/r/bitmovin/bitcodin-python?branch=master)

The bitcodin API for Python is a seamless integration with the [bitmovin cloud transcoding system](http://www.bitmovin.com). It enables the generation of MPEG-DASH and HLS content in just some minutes.

Installation 
------------

### Install with pip ###

```
sudo pip install bitcodin
```
or directly from git:

```
sudo pip install git+https://github.com/bitmovin/bitcodin-python.git
```

Usage
-----

Before you can start using the api you need to **set your API key.**

Your API key can be found in the **settings of your bitmovin user account**, as shown in the figure below.

![APIKey](https://cloudfront-prod.bitmovin.com/wp-content/uploads/2016/04/api-key.png)

An example how you can set the bitcodin API is shown in the following:

```python
import bitcodin
bitcodin.api_key = 'yourapikey'
```

**NOTE: The property names of the results returned from the api functions (e.g.: create_job()) are the same as defined in the api documentation, but are converted from camel case to snake case.**

For example:

```encoding_profile.encodingProfileId``` will be converted to ```encoding_profile.encoding_profile_id```

Example
-----
The following example demonstrates how to create a simple transcoding job:
```python

import bitcodin
bitcodin.api_key = 'yourapikey'
input_obj = bitcodin.Input(url='http://ftp.nluug.nl/pub/graphics/blender/demo/movies/Sintel.2010.720p.mkv')
input_result = bitcodin.create_input(input_obj)

video_configs = list()
video_config1 = bitcodin.VideoStreamConfig(default_stream_id=0, bitrate=4800000, profile='Main',
                                           preset='premium', height=1080, width=1920)
video_config2 = bitcodin.VideoStreamConfig(default_stream_id=0, bitrate=2400000, profile='Main',
                                           preset='premium', height=720, width=1280)
video_config3 = bitcodin.VideoStreamConfig(default_stream_id=0, bitrate=1200000, profile='Main',
                                           preset='premium', height=480, width=854)                                    
video_configs.append(video_config1)
video_configs.append(video_config2)
video_configs.append(video_config3)

audio_configs = list()
audio_config = bitcodin.AudioStreamConfig(default_stream_id=0, bitrate=192000)
audio_configs.append(audio_config)

encoding_profile = bitcodin.EncodingProfile('API Test Profile', video_configs, audio_configs)
encoding_profile_result = bitcodin.create_encoding_profile(encoding_profile)

manifests = ['mpd', 'm3u8']

job = bitcodin.Job(input_result.input_id, encoding_profile_result.encoding_profile_id, manifests)

job_result = bitcodin.create_job(job)

```


CHANGELOG.md
# bitcodin-python Changelog

## 0.4.0
* Implemented support for PlayReady and Widevine, both individually or combined

## 0.5.0
* Implemented support for HLS Encryption
* Added test for Multiple Audio Streams feature

## 1.0.0
* Implemented live stream support
* Added video_meta_data and extract_closed_captions functionality to job
* Added examples for DASH and HLS encryption
* Added additional error informations

## 1.0.1
* PEP8 style guide cosmetics

## 1.0.2
* Improved Live Stream example

## 1.0.3
* Improved Live Stream example

## 1.1.0
* Implemented vtt mpd support
* Added vtt mpd examples
* Implemented thumbnail support
* Added thumbnail examples
* Updated Live stream example

## 1.2.0
* Added support for deinterlacing jobs

## 1.2.2
* Refactored tests

## 1.2.3
* Removed BitcodinUnknownApiRequestUrlError and refactored tests 
* Updated examples

## 1.2.4
* Added audio merging feature example
* Added manifest info example

## 1.2.5
* Fixed wrong api base url

## 1.2.6
* Adjusted endpoint url

## 1.3.0
* Improved to_json() method for better debugging
* Added fairplay hls encryption job support
* Added create job skip analysis
* Adjusted examples

## 1.3.1
* Added example to show how to use start time and duration

## 1.4.0
* Improved camel case to snake case conversion (Fixed wrong conversion of lists)
* Adjusted tests to wait for jobs to finish
* Added some tests

## 1.4.1
* Fixed bug that to_json() converts BitcodinObject to dict

## 1.4.2
* Fixed wrong conversion of int and float of dictionary

## 1.4.3
* Fixed wrong conversion of int and float with python 2.7 support
* Added list_transfer_jobs() method
* Adjusted tests and example

## 1.4.4
* Changed to return object instead of boolean in transfer_job()

## 1.4.5
* Added flag create_sub_directory to outputs

## 1.4.6
* Changed preset to premium
* Adjusted encoding profile quality o examples

## 1.4.7
* Added keep aspect ratio feature and test

## 1.4.8
* Added basic authentication
* Added sftp output
* Added ftp input
* Added repr to BitcodinObject base class
* Removed testtools from setup.py

## 1.5.0
* Added transmux support
* Added transmux example

## 1.6.0
* Added get_thumbnail() method
* Thumbnail generation call now asynchronous

## 1.6.1
* Updated version names

## 1.6.2
* Added delete job method

## 1.6.5
* Fixed tests

## 1.6.6
* Updated readme

## 1.6.7
* Implemented asynchronous input generation

## 1.6.8
* Added list events method
* Added to create/delete/list subscriptions
* Added tests

## 1.6.9
* Added methods for boolean and length representations to BitcodinObject

examples/create_sprite.py
#!/usr/bin/env python

import bitcodin

# Set your API key
bitcodin.api_key = 'INSERT YOUR API KEY'

# Create an input
input_obj = bitcodin.Input(url='http://bitbucketireland.s3.amazonaws.com/Sintel-original-short.mkv')
input_result = bitcodin.create_input(input_obj)

# Create encoding profile
video_configs = list()

video_configs.append(bitcodin.VideoStreamConfig(
    default_stream_id=0,
    bitrate=4800000,
    profile='Main',
    preset='premium',
    height=1080,
    width=1920
))
video_configs.append(bitcodin.VideoStreamConfig(
    default_stream_id=0,
    bitrate=2400000,
    profile='Main',
    preset='premium',
    height=768,
    width=1024
))
video_configs.append(bitcodin.VideoStreamConfig(
    default_stream_id=0,
    bitrate=1200000,
    profile='Main',
    preset='premium',
    height=480,
    width=854
))

audio_configs = [bitcodin.AudioStreamConfig(default_stream_id=0, bitrate=192000)]

encoding_profile_obj = bitcodin.EncodingProfile('API Test Profile', video_configs, audio_configs)
encoding_profile_result = bitcodin.create_encoding_profile(encoding_profile_obj)

manifests = ['mpd', 'm3u8']

# Create a job
job = bitcodin.Job(
    input_id=input_result.input_id,
    encoding_profile_id=encoding_profile_result.encoding_profile_id,
    manifest_types=manifests
)
job_result = bitcodin.create_job(job)
print("Job created!")

# Create a sprite with a height of 240px, a width of 320 every 5 seconds from a given job
# Note: You don't have to create a new job for a sprite, you can use finished jobs too.
sprite_request = bitcodin.SpriteRequest(job_id=job_result.job_id, height=240, width=320, distance=5)
sprite = bitcodin.create_sprite(sprite_request)

print("Sprite generated!\nURL to sprite jpg: %s\n" % sprite.sprite_url)
print("Sprite generated!\nURL to sprite vtt: %s\n" % sprite.vtt_url)

